app/src/main/java/com/example/chrono/main/circuit/CircuitTimerActivity.kt

```diff
@@ -0,0 +1,163 @@
+import kotlin.math.roundToInt
+
+class CircuitTimerActivity : AppCompatActivity() {
+    private var bind: ActivityCircuitTimerBinding? = null
+
+    enum class TimerState { INIT, RUNNING, PAUSED }
+    enum class RunningState { INIT, WORK, REST }
+
+    private lateinit var countdown: CountDownTimer
+    private var timerState: TimerState = TimerState.INIT
+    private lateinit var runningState: RunningState
+
+    private var secondsLeft: Float = 0.0f
+    private var currentSet: Int = 0
+
+    private var sets: Int = 0
+    private var timeRest: Int = 0
+    private var timeWork: Int = 0
+
+//    private lateinit var audioPlayer: AsyncPlayer
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_circuit_timer)
+        bind = DataBindingUtil.setContentView(this, R.layout.activity_circuit_timer)
+
+        loadTimer(
+            GsonBuilder().create()
+                .fromJson(intent.getStringExtra("circuitObject"), CircuitObject::class.java)
+        )
+
+        // Initialize stuff perhaps
+
+        bind!!.startbutton.setOnClickListener {
+            workout()
+            timerState = TimerState.RUNNING
+            updateButtonUI()
+        }
+
+        bind!!.pausebutton.setOnClickListener {
+            countdown.cancel()
+            timerState = TimerState.PAUSED
+            updateButtonUI()
+        }
+
+        bind!!.resumebutton.setOnClickListener {
+            createTimer(secondsLeft.toInt(), true)
+            timerState = TimerState.RUNNING
+            updateButtonUI()
+        }
+
+        bind!!.closeButton.setOnClickListener {
+            // Make sure that the timer is shut down
+            countdown.cancel()
+            setResult(Activity.RESULT_CANCELED)
+            finish()
+        }
+    }
+
+    private fun createTimer(seconds: Int, wasPaused: Boolean) {
+        var time = seconds.toFloat() * 1000 + 250
+        if (wasPaused) {
+            time = secondsLeft * 1000 + 250
+        }
+        countdown = object : CountDownTimer(time.toLong(), 250) {
+            override fun onTick(p0: Long) {
+                if ((p0.toFloat().roundToInt() / 1000.0f) != secondsLeft) {
+                    secondsLeft = (p0.toFloat() / 1000.0f).roundToInt().toFloat()
+                    updateTimerUI()
+                }
+            }
+
+            override fun onFinish() {
+                if (currentSet != 0) {
+                    when (runningState) {
+                        RunningState.INIT -> {
+                            workout()
+                        }
+                        RunningState.WORK -> {
+                            rest()
+                        }
+                        RunningState.REST -> {
+                            workout()
+                        }
+                    }
+                } else {
+                    isDone()
+                }
+            }
+        }.start()
```

Comment:
Or rename func

Reply:
That's just the way the countdown timer works, I have to call .start() there because that's where the object was created. Not entirely sure what's wrong here. I have no choice but to override the onFinish and onTick logic here anyways.

Reply:
Guess I can rename it